src/plugins/spaces_oss/public/plugin.ts

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { CoreSetup, CoreStart, PluginInitializerContext, Plugin } from 'src/core/public';
+import { SpacesOssPluginSetup, SpacesOssPluginStart } from './types';
+import { SpacesApi } from './api';
+
+export class SpacesOssPlugin implements Plugin<SpacesOssPluginSetup, SpacesOssPluginStart, {}, {}> {
+  private apiRegistered = false;
+  private api?: SpacesApi;
+
+  constructor(context: PluginInitializerContext) {}
+
+  public setup({}: CoreSetup) {
+    return {
+      registerSpacesApi: (provider: Promise<SpacesApi>) => {
```

Comment:
Does this need to be wrapped in a promise? It looks like the only thing we need to create the `SpacesApi` is an instance of the `SpacesManager`, which is constructed and fully available during the setup phase. Removing the promise here would simplify the logic here a bit, and it would also allow us to guarantee that `this.api` and `this.apiRegistered` are definitively set (or intentionally unset) by the time we get to the `start` phase.

It also allows us to provide a "default" implementation for the API in the start phase (see other comment below), which effectively prevents someone from accidently trying to call `registerSpacesApi` after the `start` phase has begun.

Reply:
Well this PR was put together as preliminary work to share UI components. The `ShareSavedObjectsToSpaceFlyout` relies on start services (both for its context wrapper and for toasts), so I think we'll need to keep the promise. I should have mentioned that in the PR description, sorry about that!

Reply:
As discussed offline, I changed this in 1043483.